shell/shared/db.js

```diff
@@ -86,6 +86,46 @@ Grains = new Mongo.Collection("grains");
 //   publicId:  An id used to publicly identify this grain. Used e.g. to route incoming e-mail and
 //       web publishing. This field is initialized when first requested by the app.
 
+RoleAssignments = new Mongo.Collection("roleAssignments");
+// A role assignment allows a user to access a grain with a specified set of permissions; it is an
+// edge in the permissions sharing graph. A user's permissions for a given grain are computed as
+// the union of the permissions received through role assignments.
+//
+// Any grain for which a user has received a role assignment will show up in that user's grain list.
+// Role assignments are also used to restrict access during restoration of persisted capabilities.
+//
+// Each contains:
+//   _id: random
+//   grainId: The `_id` of the grain whose permissions are being shared.
+//   sharer: The `_id` of the user who is sharing these permissions.
+//   recipient: The `_id` of the user who receives these permissions.
+//   permissions: An array of bools representing the received permissions.
+//   roleAssignment: A Grain.ViewSharingLink.RoleAssignment that is equivalent to `permissions`, but
+//                   may contain more semantic information. In JSON format.
+//   petname: Human-readable label chosen by and only visible to the sharer.
+//   title: Human-readable title as chosen by the recipient. Used in the same places that
+//          `grain.title` is used for the grain's owner.
+//   created: Date when this role assignment was created.
+//   expires: Optional expiration date. If undefined, the role assignment does not expire.
+//   parentKey: If present, the `_id` of the role assignment key from which this was derived.
+
+RoleAssignmentKeys - new MongoCollection("sharingUrls");
+// Role assignments that are not yet bound to a single recipient. These can be used to implement
+// sharing-by-secret-URL.
+//
+// TODO(soon): Does it make sense to merge this collection with RoleAssignments?
+//
+// Each contains:
+//   _id: random
+//   sharer: The `_id` of the user who created this key.
+//   grainId: The `_id` of the grain whose permissions are being shared.
+//   permissions: An array of bools representing the received permissions.
+//   roleAssignment: A Grain.ViewSharingLink.RoleAssignment that is equivalent to `permissions`, but
+//                   may contain more semantic information. In JSON format.
+//   petname: Human-readable label chosen by and only visible to the sharer.
+//   created: Date when this key was created.
+//   expires: Optional date when this key and all role assignments derived from it should expire.
```

Comment:
I think we may want separate fields for when the key itself expires (and can no longer be exercised to grant access to new people) and when access granted via the key expires. In fact, the former is probably much more useful in practice than the latter: it lets you guard against long-term leakage of keys without needing to re-send new ones to the people you actually meant to share with.